engine/src/main/java/io/camunda/zeebe/engine/state/migration/to_8_3/ProcessInstanceByProcessDefinitionMigration.java

```diff
@@ -22,6 +24,12 @@ public String getIdentifier() {
     return getClass().getSimpleName();
   }
 
+  @Override
+  public boolean needsToRun(final ProcessingState processingState) {
+    return processingState.isEmpty(ZbColumnFamilies.PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY)
+        && !processingState.isEmpty(ZbColumnFamilies.ELEMENT_INSTANCE_KEY);
+  }
+
```

Comment:
❌ I believe this condition is wrong. In the current migration setup, it's possible that:
- a migration task was run by a leader storing it in a snapshot
- the snapshot is installed on a follower running a previous version
- the follower doesn't apply events the same way, but does build state in follower-mode
- when becoming leader (after upgrade to the newer version), the follower must still run the migration on this newly added data to the `ELEMENT_INSTANCE_KEY` CF
- it won't run the migration because there are already entries in the `PROCESS_INSTANCE_KEY_BY_DEFINITION_KEY` CF
- So not all entries in the `ELEMENT_INSTANCE_KEY` CF have a corresponding entry in the `PROCESS_INSTANCE_KEY_BY_DEFITINION_KEY` CF

Reply:
AFAIk there is no way to prevent this except for always running the migration. This is what we've been doing since we had migrations. Do you know a better check?

Reply:
Just `!processingState.isEmpty(ZbColumnFamilies.ELEMENT_INSTANCE_KEY)` should be enough right

Reply:
Oh wait, problem is that this is then run every time....

Reply:
Exactly 😅

Reply:
I'm confused why this is a problem now and for all the other older migration we've had since 1.3. Do you know?

Reply:
🤔 How about a different approach... Can we use `parentChildColumnFamily` to find all process instances?

This CF contains either:
1. 👍 `[-1, process instance key] -> [nil]` for process instances
2. 🚫 `[an element instance key, another element instance key] -> [nil]` for any other element instance

Point 1 is also the case for child process instances, as no flowScopeKey is set for the child process instance: https://github.com/camunda/zeebe/blob/013a8bcd888b5d715593eaeb16bcbae62fcad5b4/engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnStateTransitionBehavior.java#L498-L516. Which is used by the activating applier to determine the parent of the instance to create: https://github.com/camunda/zeebe/blob/013a8bcd888b5d715593eaeb16bcbae62fcad5b4/engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java#L54-L56. Which is used to set the parent of the instance to create: https://github.com/camunda/zeebe/blob/013a8bcd888b5d715593eaeb16bcbae62fcad5b4/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java#L148-L152. And finally, inserted with either `-1` or an actual element instance key: https://github.com/camunda/zeebe/blob/013a8bcd888b5d715593eaeb16bcbae62fcad5b4/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbElementInstanceState.java#L198-L201

Reply:
>I'm confused why this is a problem now and not for all the other older migration we've had since 1.3. Do you know?

Perhaps those aren't correct either. Let's focus on this one first, and then we look at the others again to make sure.

Reply:
> How about a different approach... Can we use `parentChildColumnFamily` to find all process instances?

TIL we insert a parent-child relationship for something without a parent. That's so weird 😆 I guess it could work then. We iterate with startAtKey `-1` and we should find everything without a parent, which if I understand correct would be all process instances.

Reply:
> Use the -1 in a whileEqualPrefix to find all process instances that need to be migrated. Perhaps we can also use this to know whether it must be run or not.

I don't see how it solves the migration issue. We can iterate over a different CF but the problem remains the same. What I was thinking of is we could remove the migration altogether and just use the parent-child column family. It's kind of confusing but it solves the migration issue until we have a better solution.

Reply:
> What I was thinking of is we could remove the migration altogether and just use the parent-child column family. 

This will actually be a big performance hit if there's many process instances as it requires looping over all of them, getting the element instance and checking if the process definition key matches.

Reply:
>  It can also help determine whether it should be run, if we can efficiently count all entries. It's supposed to be a 1:1 relation, so if the counts don't match we must run migration. I'm not sure how efficiently we can count though

Clever, I like it. Let's see how easy it is to find the amount of entries.

> Hmm, so you're saying it isn't a solution to remove the migration altogether? Or are you referring here to using the parent-child column family to determine whether to run the migration? Or something else?

I was referring to determine if there's running instances using the parent-child CF. It's a possibility, but I prefer the migration as it will be infinitely faster. If we would go this way we'd iterate over more element instances and have to manually get them from the state to be able to check the process definition.

Reply:
We resolved this by introducing a count (with prefix) method and comparing the entries in the two column families.